crates/core/src/pattern_compiler/compiler.rs

```diff
@@ -567,6 +567,121 @@ fn find_definition_if_exists(
+
+    let tree = parser.parse_file(src, Some(Path::new(DEFAULT_FILE_NAME)))?;
+
+    let DefsToFilenames {
+        patterns: pattern_file,
+        predicates: predicate_file,
+        functions: function_file,
+        foreign_functions: foreign_file,
+    } = defs_to_filenames(libs, parser, tree.root_node())?;
+
+    let mut traversed_stack = <Vec<String>>::new();
+
+    // gross but necessary due to running these patterns before and after each file
+    let mut stack: Vec<Tree> = if will_autowrap {
+        let before_each_file = "before_each_file()";
+        let before_tree =
+            parser.parse_file(before_each_file, Some(Path::new(DEFAULT_FILE_NAME)))?;
+        let after_each_file = "after_each_file()";
+        let after_tree = parser.parse_file(after_each_file, Some(Path::new(DEFAULT_FILE_NAME)))?;
+
+        vec![tree, before_tree, after_tree]
+    } else {
+        vec![tree]
+    };
+    while let Some(tree) = stack.pop() {
+        let root = tree.root_node();
+        let cursor = root.walk();
+
+        for n in traverse(cursor, Order::Pre).filter(|n| {
+            n.node.is_named() && (n.node.kind() == node_like || n.node.kind() == predicate_call)
+        }) {
+            let name = n
+                .child_by_field_name("name")
+                .ok_or_else(|| anyhow!("missing name of nodeLike"))?;
+            let name = name.text()?;
+            let name = name.trim().to_string();
+
+            if target_patterns.contains(&name) {
+                while let Some(e) = traversed_stack.pop() {
+                    dependents.push(e);
+                }
+            }
+            if n.node.kind() == node_like {
+                if let Some(tree) = find_child_tree_definition(
+                    &pattern_file,
+                    parser,
+                    libs,
+                    &mut traversed_stack,
+                    &name,
+                )? {
+                    stack.push(tree);
+                }
+                if let Some(tree) = find_child_tree_definition(
+                    &function_file,
+                    parser,
+                    libs,
+                    &mut traversed_stack,
+                    &name,
+                )? {
+                    stack.push(tree);
+                }
+                if let Some(tree) = find_child_tree_definition(
+                    &foreign_file,
+                    parser,
+                    libs,
+                    &mut traversed_stack,
+                    &name,
+                )? {
+                    stack.push(tree);
+                }
+            } else if n.node.kind() == predicate_call {
+                if let Some(tree) = find_child_tree_definition(
+                    &predicate_file,
+                    parser,
+                    libs,
+                    &mut traversed_stack,
+                    &name,
+                )? {
+                    stack.push(tree);
+                }
```

Comment:
Needing to repeat like this is a code smell. Instead of having a separate lookup for each kind we should be able to collapse into a main map before hand then just traverse that.